### PR TITLE
KYLIN-3942 Support multi-level json event in backend

### PR DESCRIPTION
**Pull request by hit-lacus**
# Problem
Currently real-time OLAP didn't support  multi-level json event.
For example,if I have a kafka multi-level json event like this:

### Sample Event
```json
{"country":"JAPAN","amount":13.075058425023922,"qty":8,"currency":"USD","order_time":1554801950882,"category":"ELECTRONIC","device":"Andriod","user":{"gender":"Female","id":"7a0cfa5e-bbaa-79ef-1a38-e06f02c85fcb","first_name":"unknown","age":16}}
```

### streaming_receiver.log
```
2019-04-09 09:46:09,878 ERROR [StreamingV2Cube_channel] kafka.TimedJsonStreamParser:107 : error
com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize instance of `java.lang.String` out of START_OBJECT token
at [Source: (String)"\{"country":"US","amount":14.498498222823619,"qty":1,"currency":"USD","order_time":1554803169876,"category":"Other","device":"Other","user":{"gender":"Female","id":"0736b41a-9ae7-9b4a-a124-f74436d3eb41","first_name":"unknown","age":26}}"; line: 1, column: 140] (through reference chain: java.util.HashMap["user"])
at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)
at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1342)
at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1138)
at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1092)
at com.fasterxml.jackson.databind.deser.std.StringDeserializer.deserialize(StringDeserializer.java:63)
at com.fasterxml.jackson.databind.deser.std.StringDeserializer.deserialize(StringDeserializer.java:10)
at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringKeyMap(MapDeserializer.java:527)
at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:364)
at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:29)
at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4001)
at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3030)
at org.apache.kylin.stream.source.kafka.TimedJsonStreamParser.parse(TimedJsonStreamParser.java:79)
at org.apache.kylin.stream.source.kafka.TimedJsonStreamParser.parse(TimedJsonStreamParser.java:54)
at org.apache.kylin.stream.source.kafka.consumer.KafkaConnector.nextEvent(KafkaConnector.java:110)
at org.apache.kylin.stream.core.consumer.StreamingConsumerChannel.run(StreamingConsumerChannel.java:93)
at java.lang.Thread.run(Thread.java:748)
```

# After repair

```json
{"act_type": "play", "uid": 1716, "video_id": 1786, "play_duration": 10.2966, "detail": {"Info": "dk", "Type": "abc", "location": {"province": "taiwan"}}, "video_type": "3aa17", "ts": 1555261235000, "pageview_id": "6ed3aa34d75f0", "play_times": 31}
```

image

d97bdc4e-5952-4383-a1f8-eefc06933a03

### CODE REVIEW

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by hit-lacus** [comment]
After disscussion, I use `ParseInfo`'s `columnToSourceFieldMapping` to parse multi-level json event.

企业微信截图_36c7f0aa-3f32-4c27-bf6f-0801a393e2ed

File: stream-core/src/main/java/org/apache/kylin/stream/core/source/IStreamingMessageParser.java [hunk 1]
```diff
@@ -21,5 +21,5 @@
 import org.apache.kylin.stream.core.model.StreamingMessage;
 
 public interface IStreamingMessageParser<T> {
-    StreamingMessage parse(T sourceMessage);
+    StreamingMessage parse(T sourceMessage, int partition, long offset);
```
Comment (allenma) [inline stream-core/src/main/java/org/apache/kylin/stream/core/source/IStreamingMessageParser.java]: It is better not change the interface

Comment (hit-lacus) [reply stream-core/src/main/java/org/apache/kylin/stream/core/source/IStreamingMessageParser.java]: OK

**Pull request merged**